flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/AvroToBigQuerySchemaTransform.java

```diff
@@ -0,0 +1,456 @@
+        @Nullable StandardSQLTypeName dataType;
+
+        // Handle decimal logical types by extracting precision and setting the appropriate
+        // StandardSQLTypeName.
+        // The conversions are according to
+        // https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#decimal_types
+        if (avroSchema.getProp(org.apache.avro.LogicalType.LOGICAL_TYPE_PROP) != null
+                && avroSchema.getProp(LogicalType.LOGICAL_TYPE_PROP).equals("decimal")) {
+            dataType = handleDecimalLogicalType(avroSchema, name);
+            if (dataType == StandardSQLTypeName.NUMERIC
+                    || dataType == StandardSQLTypeName.BIGNUMERIC) {
+                // The precision and scale is also set in the BigQuery Field
+                long precision =
+                        ((LogicalTypes.Decimal) avroSchema.getLogicalType()).getPrecision();
+                long scale = ((LogicalTypes.Decimal) avroSchema.getLogicalType()).getScale();
+                return Field.newBuilder(name, dataType)
+                        .setMode(Field.Mode.REQUIRED)
+                        .setPrecision(precision)
+                        .setScale(scale)
+                        .build();
+            }
+        } else {
+            dataType =
+                    Optional.ofNullable(avroSchema.getProp(LogicalType.LOGICAL_TYPE_PROP))
+                            .map(LOGICAL_AVRO_TYPES_TO_BQ_TYPES::get)
+                            .orElse(AVRO_TYPES_TO_BQ_TYPES.get(avroSchema.getType()));
+        }
+        if (dataType == null) {
+            throw new UnsupportedOperationException(
+                    getErrorMessage(avroSchema.getType().toString(), name));
+        }
+
+        return Field.newBuilder(name, dataType).setMode(Field.Mode.REQUIRED).build();
+    }
+
+    /**
+     * Helper method to handle Avro decimal logical types by determining the appropriate BigQuery
+     * data type based on the precision of the decimal.
+     *
+     * @param avroSchema the Avro schema of the decimal field
+     * @param name the name of the field
+     */
+    private static StandardSQLTypeName handleDecimalLogicalType(
+            org.apache.avro.Schema avroSchema, String name) {
+        long precision = validatePrecisionAndScale(avroSchema, name);
+        long scale = ((LogicalTypes.Decimal) avroSchema.getLogicalType()).getScale();
+
+        if (precision > 0 && precision <= (scale + 29) && scale <= 9) {
+            return StandardSQLTypeName.NUMERIC;
+        } else if (precision > 0 && precision <= (scale + 38) && scale <= 38) {
+            return StandardSQLTypeName.BIGNUMERIC;
+        }
+        LOG.warn(
+                "The precision {} and scale {} of decimal field {} is not supported by BigQuery. Converting the field to type STRING.",
+                precision,
+                scale,
+                name);
+        return StandardSQLTypeName.STRING;
```

Comment:
Is this how other connectors (beam, spark) handle it?

Reply:
The other connectors throw exception for these cases. The conversion rule applied here is based out of default [rules](https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-avro#logical_types) for loading avro to bigquery data in BQ.

Reply:
In that case, will the serializer also convert those types to strings? If serializer throws an error, then allowing here but failing later in inconsistent behavior. We should serialize accordingly.

Reply:
Sure, I rechecked with the serializer, error is thrown there. Hence, the same behaviour (throwing error) has been done here.